src/appleseed-max-impl/appleseedoslplugin/oslparamdlg.cpp

```diff
@@ -94,6 +94,88 @@ namespace
             }
         }
     };
+
+    void adjust_group_heights(PageGroup* page)
+    {
+        for (auto child : page->m_children)
+            adjust_group_heights(child);
+
+        page->m_height += 12;                                   // extra space for the contols
+        if (page->m_parent != nullptr)
+        {
+            page->m_parent->m_height += page->m_height;
+            page->m_parent->m_height += 3;                      // spacing between the groups
+        }
+    }
+
+    void adjust_group_positions(PageGroup* page)
+    {
+        int y_pos = page->m_y_pos + page->m_height;
+        for (auto it = page->m_children.rbegin(); it != page->m_children.rend(); ++it)
+        {
+            PageGroup* child = *it;
+            child->m_y_pos = y_pos - child->m_height - 3;       // spacing between the groups
+            y_pos = child->m_y_pos;
+
+            child->m_width = page->m_width - 8;
+            child->m_x_pos = page->m_x_pos + 4;
+        }
+
+        for (auto child : page->m_children)
+            adjust_group_positions(child);
+    }
+
+    int build_page_tree(const std::vector<OSLParamInfo>& shader_params, PageGroupMap& pages)
+    {
+        // Calculate height of each group based on controls.
+        for (auto& param : shader_params)
+        {
+            if (param.m_page != "" &&
+                param.m_widget != "null" &&
+                !param.m_max_hidden_attr)
+            {
+                const int param_height = 
+                    param.m_max_param.m_param_type == MaxParam::IntMapper ||
+                    param.m_max_param.m_param_type == MaxParam::StringPopup ? 15 : 12;
+
+                pages[param.m_page].m_height += param_height;
+            }
+        }
+        
+        // Build page tree.
+        PageGroup root_page;
+        root_page.m_width = 217;
+        std::vector<PageGroup*> page_tree;
+        for (auto& page : pages)
+        {
+            auto parent_name = page.first;
+            size_t dot_pos = parent_name.find_last_of('.');
+            if (dot_pos == std::string::npos)
+            {
+                page.second.m_short_name = page.first;
+                page_tree.push_back(&page.second);
+                page.second.m_parent = &root_page;
```

Comment:
You're taking a reference to a local variable. Is that OK?

Reply:
I should reset m_parent to nullptr before returning from function probably.